### DISCUSSION TITLE
Is it safe to share DB connection between fibers?

### DISCUSSION THREAD

**Discussion by edzhelyov** [ioquatix/ioquatix#12] [General]
Hi Samuel,

I'm trying to get Falcon working with parts of our Rails app. In regards to ActiveRecord, it uses a connection pool and each connection is checkout by a thread, because fibers are running in a single thread this means they will reuse the same database connection. Is it safe to do so?

**Comment by edzhelyov** [comment]
--nevermind-- I had a race condition in the test file where each fiber checkout a new connection. This was due to the fact Mutex is not Fiber aware and wasn't blocking the queue. When I preload the AR's connection before running the fibers I can reproduce the shared single connection.

**Reply by ioquatix** [reply]
Mutex in Ruby 3.x is fiber aware.

**Comment by AmyYummy** [comment]
I think it's possible, the [fiber optic](www.baudcom.cn) guy passed by.

**Comment by ioquatix** [comment]
Just following up, https://github.com/rails/rails/issues/42271 is a discussion relating to this problem and the current proposed solutions. Some parts are being merged which is good news.

**Reply by kinnalru** [reply]
I'am testing production IO-bound  project with falcon. Just upgrade to rails-edge and add isolation level :fiber in application.rb. It is amazing concurrency boost! 
Simple production test: ab with concurrency 100, project writes to DB through AR and uploads file(~5kb) to Minio through AWS::SDK::S3. Threre RPS and cpu core load in cells

| Server    | 1x1            | 1x10     | 2x1            | 2x10       | 5x1            | 5x10      | 10x1                  |
| --------- | -------------- | -------- | -------------- | ---------- | -------------- | --------- | --------------------- |
| Puma      | 8 1x5%         | 85 1x65% | 10 2x5%        | 165 2x70%  | 40 5x7%        | 369 5x70% | 89   10x8%            |
| Passenger | 6 1x5%         |          | 15 2x5%        |            | 51 5x5%        |           | 80 10x7%              |
| Falcon    | **134** 1x100% |          | **238** 2x100% |  | **445** 5x100% |           | 470 10x50% - underload |

**Reply by ioquatix** [reply]
@kinnalru can you explain a little bit more what the numbers mean? What is 2x1? Is the number given in bold the requests per second? What is 2x100%? It means both cpu cores are max utilization?

**Reply by kinnalru** [reply]
@ioquatix yep
"AxB" in headers is process and threads - 2x5 - 2 processes with 5 threads in each.
"R CxP%" in cells - requests per seconds and core/cpu utilization - 6 1x5% means "6 rps and 1 core loaded by 5%". Empty cell - when no threaded capabilities (like free Passenger) or (in case of Falcon) when I unable to control threads count. I'am not using Falcon hybrid mode - just forked.

Bold is just best values.

All tests made under stable load with Apache Benchmark with concurrency 100.
